### ISSUE TITLE
How can I generate the metrics that can be obtained after k-fold cross-validation training?

### ISSUE THREAD

**Issue by enescicek** [ultralytics/ultralytics#23424]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

@glenn-jocher 
After completing the YOLO k-fold cross-validation training, how can I obtain important metrics such as IOU score, Dice score, average performance, variance between folds, and more?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @enescicek, thank you for your interest in Ultralytics 🚀! This is an automated response to help you get unstuck quickly—an Ultralytics engineer will also assist soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it (especially helpful for k-fold setups).  
If this is a custom training ❓ Question (like k-fold cross-validation metrics), please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/). 🙏

To help the team answer precisely, please add:
-Your exact k-fold workflow (how you split folds, how you train each fold, and how you evaluate)
-The command(s) you run (CLI) or the training script (Python)
-The `ultralytics` version and environment details
-Where you expect metrics to appear (console output, `runs/` logs, exported files, etc.)
-Example outputs/log snippets from one fold and how you’re trying to aggregate them 📊

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by glenn-jocher** [comment]
Ultralytics doesn’t currently provide a built-in “k-fold aggregator”, so the usual workflow is to run `model.val()` once per fold (each fold’s `best.pt`), collect the returned scalar metrics from `results.results_dict`, and then compute your mean/variance across folds; you can also get per-class tables via `results.summary()` as shown in the [metrics reference](https://docs.ultralytics.com/reference/utils/metrics/).

```python
from ultralytics import YOLO
import numpy as np

fold_weights = [
    "runs/segment/fold0/weights/best.pt",
    "runs/segment/fold1/weights/best.pt",
    "runs/segment/fold2/weights/best.pt",
    "runs/segment/fold3/weights/best.pt",
    "runs/segment/fold4/weights/best.pt",
]

fold_metrics = []
for w in fold_weights:
    results = YOLO(w).val(data="your_dataset.yaml")  # optionally add split=... if you use a custom split
    fold_metrics.append(results.results_dict)

keys = fold_metrics[0].keys()
mean = {k: float(np.mean([m[k] for m in fold_metrics])) for k in keys}
std = {k: float(np.std([m[k] for m in fold_metrics], ddof=1)) for k in keys}
print("mean:", mean)
print("std:", std)
```

For Dice / mean-IoU specifically: those are not reported as default validation outputs (Ultralytics focuses on mAP-based metrics), so you’d need to compute them yourself from predicted masks vs GT masks (often easiest with TorchMetrics) after each fold’s validation/prediction run.

**Comment by enescicek** [comment]
@glenn-jocher 

How can I obtain the IOU score, Dice score, 95% confidence interval, and their graphs after the yolov8 segmentation training?

**Comment by glenn-jocher** [comment]
After segmentation training you can generate the built-in box+mask PR/F1/metric-confidence plots by running validation with `plots=True`, and you can also pull the raw curve arrays via `metrics.curves_results` (see the implementations in [SegmentMetrics.process()](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/utils/metrics.py#L1216-L1241) and [Metric.curves_results](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/utils/metrics.py#L1000-L1007)), plus per-class tables via `metrics.summary()` ([SegmentMetrics.summary()](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/utils/metrics.py#L1288-L1314)).  

```python
from ultralytics import YOLO

model = YOLO("runs/segment/train/weights/best.pt")
metrics = model.val(data="your_dataset.yaml", plots=True)
curves = metrics.curves_results
per_class = metrics.summary(decimals=4)
```

IoU/Dice and 95% confidence intervals aren’t produced by the built-in validator, so you’ll need to compute per-image IoU/Dice from your binary predicted/GT masks and then bootstrap those scores to get a 95% CI (and plot them however you like, e.g. histogram/violin).
